### ISSUE TITLE
Bug accessing config variable from within the shell directive in a module

### ISSUE THREAD

**Issue by ImNotaGit** [snakemake/snakemake#1688]
**Snakemake version**
v7.8.0

**Describe the bug**
In a module, when the global config is overwritten with a `config` directive, accessing config variables from within the `shell` directive of rules in the module will lead to error, unless config variables of the same names are also present in the global config.

Besides, dry-run log contains mistakes not reflecting what's going on during the actual run.

**Minimal example **
Working dir structure:
```
├── config
│   └── config.yaml
├── Snakefile
└── m.smk
```
`Snakefile`:
```
configfile: "config/config.yaml"

rule all:
    input: "a.out", "m/a.out"

rule a:
    output: "a.out"
    shell: "echo {config[x]} > {output}"

module m:
    snakefile: "m.smk"
    prefix: "m"
    config: config["m"]

use rule * from m as m_*
```
`m.smk`:
```
configfile: "config/config_m.yaml"

print(config["y"])

rule a:
    output: "a.out"
    shell: "echo {config[x]} > {output}; echo {config[y]} >> {output}"
```
`config/config.yaml`:
```
x: x_global

m:
  x: x_module
  y: y_module
```
Trying to run this example will correctly print the `y` value in `config["m"]` (i.e. `"y_module"`), but the pipeline won't run with error:
```
RuleException in line 5 of .../m.smk:
NameError: The name 'y' is unknown in this context. Please make sure that you defined that variable. Also note that braces not used for variable access have to be escaped by repeating them, i.e. {{print $1}}
```
However, if modifying `config/config.yaml` to:
```
x: x_global
y: y_global

m:
  x: x_module
  y: y_module
```
This will run without error, with all output files generated with correct contents, esp. `m/a.out`:
```
x_module
y_module
```
However, note that if doing a dry-run, the logs will not give the correct shell commands used, i.e. it does not match what is actually executed during the actual run. Dry-run logs below, note the line `echo x_global > m/a.out; echo y_global >> m/a.out`:
```
y_module
Building DAG of jobs...
Job stats:
job      count    min threads    max threads
-----  -------  -------------  -------------
a            1              1              1
all          1              1              1
m_a          1              1              1
total        3              1              1


[Mon May 30 10:39:09 2022]
rule a:
    output: a.out
    jobid: 1
    reason: Missing output files: a.out; Code has changed since last execution
    resources: tmpdir=/tmp

echo x_global > a.out

[Mon May 30 10:39:09 2022]
rule m_a:
    output: m/a.out
    jobid: 2
    reason: Missing output files: m/a.out; Code has changed since last execution
    resources: tmpdir=/tmp

echo x_global > m/a.out; echo y_global >> m/a.out

[Mon May 30 10:39:09 2022]
localrule all:
    input: a.out, m/a.out
    jobid: 0
    reason: Input files updated by another job: m/a.out, a.out
    resources: tmpdir=/tmp

Job stats:
job      count    min threads    max threads
-----  -------  -------------  -------------
a            1              1              1
all          1              1              1
m_a          1              1              1
total        3              1              1


This was a dry-run (flag -n). The order of jobs does not reflect the order of execution.
```

**Comment by LukaP-BB** [comment]
x and y are not defined, you should put them in quotes, as you would access a value inside a dictionary 
```py
rule a:
    output: "a.out"
    shell: "echo {config['x']} > {output}; echo {config['y']} >> {output}"
```

**Comment by ImNotaGit** [comment]
> It isn't the prettiest solution, but if you access the config from something else, like input or params, it appears to work. So changing your _m.smk_ file to
> 
> ```
> rule a:
>     output: "a.out"
>     params:
>         x = config["x"],
>         y = config["y"]
>     shell: "echo {params.x} > {output}; echo {params.y} >> {output}"
> ```
> 
> It is a pain because it just adds lines of nearly-redundant code, but at least gets around this issue. Presumably config statements are expanded in the shell at a different time to when the global/module config changes?

@ASLeonard Yes thanks, indeed as noted the issue happens only within the shell directive, and the solution you provided would be a temporary workaround. This issue should still be fixed as the behavior is inconsistent depending on whether a variable of the same name exist in the global config, and can be very confusing and tricky to debug for someone who encountered this.

**Comment by ImNotaGit** [comment]
> x and y are not defined, you should put them in quotes, as you would access a value inside a dictionary
> 
> `m.smk` :
> 
> ```python
> rule a:
>     output: "a.out"
>     shell: "echo {config['x']} > {output}; echo {config['y']} >> {output}"
> ```

@LukaP-BB Thanks but I'm afraid this is not the issue. Actually according to snakemake documentation (https://snakemake.readthedocs.io/en/stable/snakefiles/configuration.html):
>For adding config placeholders into a shell command, Python string formatting syntax requires you to leave out the quotes around the key name, like so:
>```
>shell:
>    "mycommand {config[foo]} ..."
>```

Anyway I tried including the quotes, as expected it did not fix the issue.

**Comment by LukaP-BB** [comment]
Oh, right, sorry for the misleading answer, I was pretty sure I did that in one of my pipelines but after verification I always resort to ASLeonard way of doing

**Comment by sreichl** [comment]
Hi, I have just encountered this and do not want to rewrite all my workflows so that the config variables come from parameters (a step in the wrong direction for me). Has this been addressed or does a workaround exist?

Thanks & Cheers,
Stephan
